scipy/optimize/optimize.py

```diff
@@ -744,10 +744,13 @@ def _minimize_neldermead(func, x0, args=(), callback=None,
         sim = np.clip(sim, lower_bound, upper_bound)
 
     one2np1 = list(range(1, N + 1))
-    fsim = np.empty((N + 1,), float)
 
-    for k in range(N + 1):
-        fsim[k] = func(sim[k])
```

Comment:
Stepping back for a moment, do we need to change this? The documentation states `func` should be a scalar-valued function; so if NumPy is changing what is acceptable as scalar-valued then I don't see why it's SciPy's business to change that.

That is, unless other minimizers explicitly handle length-1 1d arrays. In which case we should probably deprecate them together.

Reply:
In most places, scipy treats floats correctly, but historically, the optimization routines don't. They allow the user the return and size-1 array as a "float" and relied on numpy to handle the conversion implicitly. The changes in this PR make the conversion explicit in scipy. An alternative would indeed be to adapt the test functions in scipy instead return true scalars, and leave it to the users to change their functions to avoid an error/deprecation warning.